# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # TODO: Implement Function
    
    source_id_text = [[source_vocab_to_int[w] for w in s.split()] for s in source_text.split('\n')]
    target_id_text = [[target_vocab_to_int[w] for w in s.split() + ['<EOS>']]  for s in target_text.split('\n')] 
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.0


D:\Anaconda\envs\tensorflow\lib\site-packages\ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [35]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    # TODO: Implement Function
    inputs = tf.placeholder(shape=(None, None), dtype=tf.int32, name='input')
    targets = tf.placeholder(shape=(None, None), dtype=tf.int32, name='target')
    learning_rate = tf.placeholder(dtype=tf.float32)
    keep_prob = tf.placeholder(dtype=tf.float32, name='keep_prob')
    targets_seq_length = tf.placeholder(shape=(None,), dtype=tf.int32, name='target_sequence_length')
    max_target_seq = tf.reduce_max(targets_seq_length, name="max_target_len")
    input_seq_length = tf.placeholder(shape=(None,), dtype=tf.int32, name="source_sequence_length")
    
    return inputs, targets, learning_rate, keep_prob, targets_seq_length, max_target_seq, input_seq_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "D:\\Anaconda\\envs\\tensorflow\\lib\\runpy.py", line 193, in _run_module_as_main\n    "__main__", mod_spec)', 'File "D:\\Anaconda\\envs\\tensorflow\\lib\\runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "D:\\Anaconda\\envs\\tensorflow\\lib\\site-packages\\ipykernel_launcher.py", line 16, in <module>\n    app.launch_new_instance()', 'File "D:\\Anaconda\\envs\\tensorflow\\lib\\site-packages\\traitlets\\config\\application.py", line 658, in launch_instance\n    app.start()', 'File "D:\\Anaconda\\envs\\tensorflow\\lib\\site-packages\\ipykernel\\kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File "D:\\Anaconda\\envs\\tensorflow\\lib\\site-packages\\zmq\\e

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [19]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    dec_input = tf.concat(
        [tf.fill(
            [batch_size, 1],
            target_vocab_to_int['<GO>']),
         ending],
        1)
    return dec_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [21]:
from imp import reload
reload(tests)

def dropout_lstm_cell(rnn_size, keep_prob):
    cell = tf.contrib.rnn.LSTMCell(
        rnn_size,
        initializer=tf.random_uniform_initializer(-0.1,0.1,seed=2))
    cell = tf.contrib.rnn.DropoutWrapper(cell, output_keep_prob = keep_prob)
    return cell

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    embedding = tf.contrib.layers.embed_sequence(
        rnn_inputs,
        source_vocab_size,
        encoding_embedding_size)   

    
    enc_cell = tf.contrib.rnn.MultiRNNCell(
        [dropout_lstm_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    enc_output, enc_state = tf.nn.dynamic_rnn(
        enc_cell, embedding, sequence_length=source_sequence_length, dtype=tf.float32)
    return enc_output, enc_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [29]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    
    training_helper = tf.contrib.seq2seq.TrainingHelper(
        inputs=dec_embed_input,
        sequence_length=target_sequence_length,
        time_major=False)
    
    training_decoder = tf.contrib.seq2seq.BasicDecoder(
        dec_cell,
        training_helper,
        encoder_state,
        output_layer)
    
    dec_outputs, _, _ = tf.contrib.seq2seq.dynamic_decode(
        training_decoder,
        impute_finished=True,
        maximum_iterations=max_summary_length)
    return dec_outputs



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [28]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    
    start_tokens = tf.tile(
        tf.constant([start_of_sequence_id], dtype=tf.int32),
        [batch_size],
        name='start_tokens')
    
    inference_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(
        dec_embeddings,
        start_tokens,
        end_of_sequence_id)
    
    inference_decoder = tf.contrib.seq2seq.BasicDecoder(
        dec_cell,
        inference_helper,
        encoder_state,
        output_layer)
    
    inf_dec_output, _, _ = tf.contrib.seq2seq.dynamic_decode(
        inference_decoder,
        impute_finished=True,
        maximum_iterations=max_target_sequence_length)
    return inf_dec_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [31]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    
    dec_embeddings = tf.Variable(
        tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
    
    dec_cell = tf.contrib.rnn.MultiRNNCell(
        [dropout_lstm_cell(rnn_size, keep_prob) for _ in range(num_layers)])
    
    output_layer = Dense(
        target_vocab_size,
        kernel_initializer=tf.truncated_normal_initializer(mean=0.0,stddev=0.1))
    
    with tf.variable_scope("decode"):
        train_dec_output = decoding_layer_train(
            encoder_state,
            dec_cell,
            dec_embed_input,
            target_sequence_length,
            max_target_sequence_length,
            output_layer,
            keep_prob)
        
    with tf.variable_scope("decode",reuse=True):
        inf_dec_output = decoding_layer_infer(
            encoder_state,
            dec_cell,
            dec_embeddings,
            target_vocab_to_int['<GO>'],
            target_vocab_to_int['<EOS>'],
            max_target_sequence_length,
            target_vocab_size,
            output_layer,
            batch_size,
            keep_prob)
    
    return train_dec_output, inf_dec_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [32]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # TODO: Implement Function
    _, enc_state = encoding_layer(
        input_data,
        rnn_size,
        num_layers,
        keep_prob,
        source_sequence_length,
        source_vocab_size,
        enc_embedding_size)
    
    dec_input = process_decoder_input(
        target_data, target_vocab_to_int, batch_size)
    
    train_dec_output, inference_dec_output = decoding_layer(
        dec_input,
        enc_state,
        target_sequence_length,
        max_target_sentence_length,
        rnn_size,
        num_layers,
        target_vocab_to_int,
        target_vocab_size,
        batch_size,
        keep_prob,
        dec_embedding_size)
    return train_dec_output, inference_dec_output


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [33]:
# Number of Epochs
epochs = 50
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 64
# Number of Layers
num_layers = 2
# Embedding Size
encoding_embedding_size = 16
decoding_embedding_size = 16
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.8
display_step = 32

### Build the Graph
Build the graph using the neural network you implemented.

In [36]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [37]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [38]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   32/538 - Train Accuracy: 0.2811, Validation Accuracy: 0.3453, Loss: 3.7695
Epoch   0 Batch   64/538 - Train Accuracy: 0.3199, Validation Accuracy: 0.3509, Loss: 3.2100
Epoch   0 Batch   96/538 - Train Accuracy: 0.3536, Validation Accuracy: 0.3828, Loss: 3.0186
Epoch   0 Batch  128/538 - Train Accuracy: 0.3536, Validation Accuracy: 0.3842, Loss: 2.9257
Epoch   0 Batch  160/538 - Train Accuracy: 0.3571, Validation Accuracy: 0.4016, Loss: 2.8231
Epoch   0 Batch  192/538 - Train Accuracy: 0.3862, Validation Accuracy: 0.4238, Loss: 2.7029
Epoch   0 Batch  224/538 - Train Accuracy: 0.3805, Validation Accuracy: 0.4476, Loss: 2.7670
Epoch   0 Batch  256/538 - Train Accuracy: 0.4012, Validation Accuracy: 0.4592, Loss: 2.6427
Epoch   0 Batch  288/538 - Train Accuracy: 0.4104, Validation Accuracy: 0.4727, Loss: 2.6325
Epoch   0 Batch  320/538 - Train Accuracy: 0.4462, Validation Accuracy: 0.4798, Loss: 2.4443
Epoch   0 Batch  352/538 - Train Accuracy: 0.4559, Validation Accuracy

Epoch   5 Batch  320/538 - Train Accuracy: 0.6419, Validation Accuracy: 0.6559, Loss: 0.5412
Epoch   5 Batch  352/538 - Train Accuracy: 0.6592, Validation Accuracy: 0.6440, Loss: 0.5558
Epoch   5 Batch  384/538 - Train Accuracy: 0.6509, Validation Accuracy: 0.6467, Loss: 0.5366
Epoch   5 Batch  416/538 - Train Accuracy: 0.6808, Validation Accuracy: 0.6511, Loss: 0.5245
Epoch   5 Batch  448/538 - Train Accuracy: 0.6514, Validation Accuracy: 0.6522, Loss: 0.5246
Epoch   5 Batch  480/538 - Train Accuracy: 0.6514, Validation Accuracy: 0.6520, Loss: 0.5331
Epoch   5 Batch  512/538 - Train Accuracy: 0.6600, Validation Accuracy: 0.6493, Loss: 0.5286
Epoch   6 Batch   32/538 - Train Accuracy: 0.6337, Validation Accuracy: 0.6523, Loss: 0.5199
Epoch   6 Batch   64/538 - Train Accuracy: 0.6468, Validation Accuracy: 0.6571, Loss: 0.5051
Epoch   6 Batch   96/538 - Train Accuracy: 0.6626, Validation Accuracy: 0.6550, Loss: 0.4868
Epoch   6 Batch  128/538 - Train Accuracy: 0.6603, Validation Accuracy

Epoch  11 Batch   96/538 - Train Accuracy: 0.7533, Validation Accuracy: 0.7360, Loss: 0.3162
Epoch  11 Batch  128/538 - Train Accuracy: 0.7474, Validation Accuracy: 0.7333, Loss: 0.3380
Epoch  11 Batch  160/538 - Train Accuracy: 0.7214, Validation Accuracy: 0.7308, Loss: 0.3301
Epoch  11 Batch  192/538 - Train Accuracy: 0.7134, Validation Accuracy: 0.7431, Loss: 0.3311
Epoch  11 Batch  224/538 - Train Accuracy: 0.6957, Validation Accuracy: 0.7370, Loss: 0.3431
Epoch  11 Batch  256/538 - Train Accuracy: 0.7357, Validation Accuracy: 0.7377, Loss: 0.3473
Epoch  11 Batch  288/538 - Train Accuracy: 0.7328, Validation Accuracy: 0.7395, Loss: 0.3466
Epoch  11 Batch  320/538 - Train Accuracy: 0.7251, Validation Accuracy: 0.7479, Loss: 0.3284
Epoch  11 Batch  352/538 - Train Accuracy: 0.7407, Validation Accuracy: 0.7388, Loss: 0.3509
Epoch  11 Batch  384/538 - Train Accuracy: 0.7641, Validation Accuracy: 0.7459, Loss: 0.3119
Epoch  11 Batch  416/538 - Train Accuracy: 0.7660, Validation Accuracy

Epoch  16 Batch  384/538 - Train Accuracy: 0.8110, Validation Accuracy: 0.7766, Loss: 0.2278
Epoch  16 Batch  416/538 - Train Accuracy: 0.8081, Validation Accuracy: 0.7777, Loss: 0.2326
Epoch  16 Batch  448/538 - Train Accuracy: 0.8062, Validation Accuracy: 0.7900, Loss: 0.2335
Epoch  16 Batch  480/538 - Train Accuracy: 0.8237, Validation Accuracy: 0.7860, Loss: 0.2438
Epoch  16 Batch  512/538 - Train Accuracy: 0.8110, Validation Accuracy: 0.7946, Loss: 0.2331
Epoch  17 Batch   32/538 - Train Accuracy: 0.7879, Validation Accuracy: 0.7903, Loss: 0.2271
Epoch  17 Batch   64/538 - Train Accuracy: 0.8112, Validation Accuracy: 0.7903, Loss: 0.2268
Epoch  17 Batch   96/538 - Train Accuracy: 0.8233, Validation Accuracy: 0.7898, Loss: 0.2223
Epoch  17 Batch  128/538 - Train Accuracy: 0.8358, Validation Accuracy: 0.7898, Loss: 0.2247
Epoch  17 Batch  160/538 - Train Accuracy: 0.7736, Validation Accuracy: 0.7860, Loss: 0.2328
Epoch  17 Batch  192/538 - Train Accuracy: 0.7902, Validation Accuracy

Epoch  22 Batch  160/538 - Train Accuracy: 0.8093, Validation Accuracy: 0.8288, Loss: 0.1591
Epoch  22 Batch  192/538 - Train Accuracy: 0.8391, Validation Accuracy: 0.8377, Loss: 0.1615
Epoch  22 Batch  224/538 - Train Accuracy: 0.8414, Validation Accuracy: 0.8263, Loss: 0.1747
Epoch  22 Batch  256/538 - Train Accuracy: 0.8326, Validation Accuracy: 0.8516, Loss: 0.1728
Epoch  22 Batch  288/538 - Train Accuracy: 0.8145, Validation Accuracy: 0.8526, Loss: 0.1750
Epoch  22 Batch  320/538 - Train Accuracy: 0.8464, Validation Accuracy: 0.8512, Loss: 0.1622
Epoch  22 Batch  352/538 - Train Accuracy: 0.8359, Validation Accuracy: 0.8388, Loss: 0.1849
Epoch  22 Batch  384/538 - Train Accuracy: 0.8521, Validation Accuracy: 0.8320, Loss: 0.1492
Epoch  22 Batch  416/538 - Train Accuracy: 0.8462, Validation Accuracy: 0.8422, Loss: 0.1619
Epoch  22 Batch  448/538 - Train Accuracy: 0.8447, Validation Accuracy: 0.8485, Loss: 0.1557
Epoch  22 Batch  480/538 - Train Accuracy: 0.8629, Validation Accuracy

Epoch  27 Batch  448/538 - Train Accuracy: 0.9062, Validation Accuracy: 0.8990, Loss: 0.1184
Epoch  27 Batch  480/538 - Train Accuracy: 0.9172, Validation Accuracy: 0.8958, Loss: 0.1149
Epoch  27 Batch  512/538 - Train Accuracy: 0.9087, Validation Accuracy: 0.8945, Loss: 0.1187
Epoch  28 Batch   32/538 - Train Accuracy: 0.9070, Validation Accuracy: 0.8908, Loss: 0.1113
Epoch  28 Batch   64/538 - Train Accuracy: 0.8903, Validation Accuracy: 0.8904, Loss: 0.1211
Epoch  28 Batch   96/538 - Train Accuracy: 0.9315, Validation Accuracy: 0.8970, Loss: 0.1128
Epoch  28 Batch  128/538 - Train Accuracy: 0.9131, Validation Accuracy: 0.9050, Loss: 0.1221
Epoch  28 Batch  160/538 - Train Accuracy: 0.9057, Validation Accuracy: 0.8990, Loss: 0.1111
Epoch  28 Batch  192/538 - Train Accuracy: 0.9245, Validation Accuracy: 0.8862, Loss: 0.1136
Epoch  28 Batch  224/538 - Train Accuracy: 0.9205, Validation Accuracy: 0.8871, Loss: 0.1294
Epoch  28 Batch  256/538 - Train Accuracy: 0.8982, Validation Accuracy

Epoch  33 Batch  224/538 - Train Accuracy: 0.9240, Validation Accuracy: 0.9103, Loss: 0.0890
Epoch  33 Batch  256/538 - Train Accuracy: 0.9131, Validation Accuracy: 0.9153, Loss: 0.0818
Epoch  33 Batch  288/538 - Train Accuracy: 0.9350, Validation Accuracy: 0.9086, Loss: 0.0874
Epoch  33 Batch  320/538 - Train Accuracy: 0.9275, Validation Accuracy: 0.9121, Loss: 0.0844
Epoch  33 Batch  352/538 - Train Accuracy: 0.9077, Validation Accuracy: 0.9142, Loss: 0.0984
Epoch  33 Batch  384/538 - Train Accuracy: 0.9195, Validation Accuracy: 0.9196, Loss: 0.0792
Epoch  33 Batch  416/538 - Train Accuracy: 0.9205, Validation Accuracy: 0.9100, Loss: 0.0795
Epoch  33 Batch  448/538 - Train Accuracy: 0.9315, Validation Accuracy: 0.9167, Loss: 0.0707
Epoch  33 Batch  480/538 - Train Accuracy: 0.9368, Validation Accuracy: 0.9107, Loss: 0.0817
Epoch  33 Batch  512/538 - Train Accuracy: 0.9249, Validation Accuracy: 0.9130, Loss: 0.0881
Epoch  34 Batch   32/538 - Train Accuracy: 0.9202, Validation Accuracy

Epoch  38 Batch  512/538 - Train Accuracy: 0.9122, Validation Accuracy: 0.9281, Loss: 0.0701
Epoch  39 Batch   32/538 - Train Accuracy: 0.9302, Validation Accuracy: 0.9224, Loss: 0.0578
Epoch  39 Batch   64/538 - Train Accuracy: 0.9351, Validation Accuracy: 0.9176, Loss: 0.0678
Epoch  39 Batch   96/538 - Train Accuracy: 0.9472, Validation Accuracy: 0.9231, Loss: 0.0554
Epoch  39 Batch  128/538 - Train Accuracy: 0.9312, Validation Accuracy: 0.9322, Loss: 0.0622
Epoch  39 Batch  160/538 - Train Accuracy: 0.9312, Validation Accuracy: 0.9281, Loss: 0.0605
Epoch  39 Batch  192/538 - Train Accuracy: 0.9390, Validation Accuracy: 0.9210, Loss: 0.0605
Epoch  39 Batch  224/538 - Train Accuracy: 0.9350, Validation Accuracy: 0.9190, Loss: 0.0731
Epoch  39 Batch  256/538 - Train Accuracy: 0.9277, Validation Accuracy: 0.9338, Loss: 0.0708
Epoch  39 Batch  288/538 - Train Accuracy: 0.9426, Validation Accuracy: 0.9231, Loss: 0.0621
Epoch  39 Batch  320/538 - Train Accuracy: 0.9373, Validation Accuracy

Epoch  44 Batch  288/538 - Train Accuracy: 0.9514, Validation Accuracy: 0.9419, Loss: 0.0570
Epoch  44 Batch  320/538 - Train Accuracy: 0.9408, Validation Accuracy: 0.9395, Loss: 0.0584
Epoch  44 Batch  352/538 - Train Accuracy: 0.9111, Validation Accuracy: 0.9462, Loss: 0.0790
Epoch  44 Batch  384/538 - Train Accuracy: 0.9304, Validation Accuracy: 0.9288, Loss: 0.0539
Epoch  44 Batch  416/538 - Train Accuracy: 0.9375, Validation Accuracy: 0.9425, Loss: 0.0643
Epoch  44 Batch  448/538 - Train Accuracy: 0.9312, Validation Accuracy: 0.9441, Loss: 0.0536
Epoch  44 Batch  480/538 - Train Accuracy: 0.9531, Validation Accuracy: 0.9389, Loss: 0.0563
Epoch  44 Batch  512/538 - Train Accuracy: 0.9382, Validation Accuracy: 0.9400, Loss: 0.0591
Epoch  45 Batch   32/538 - Train Accuracy: 0.9513, Validation Accuracy: 0.9368, Loss: 0.0489
Epoch  45 Batch   64/538 - Train Accuracy: 0.9328, Validation Accuracy: 0.9325, Loss: 0.0563
Epoch  45 Batch   96/538 - Train Accuracy: 0.9593, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [39]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [40]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [ ]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """      
    return [vocab_to_int[w] for w in sentence.lower.split()]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

## Translate
This will translate `translate_sentence` from English to French.

In [ ]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.